equational-reasoning.md

```diff
@@ -0,0 +1,25 @@
+Equational reasoning
+====================
+
+**Tags**: *substitution model*, *referential transparency*, *purity*, *side effect*
+
+In terms of programming we can say that a function definition defines the scope where different expressions are evaluated in order to eventually return a result. This result defines the return type of that function.
+
+For developers it seems very convenient to be able to `reason` about a function (i.e. understand what the function does ) just by checking the value that it returns. This is to say that a function's return value represents the transformation of its internal state by potentially using other arguments as its parameters.
```

Comment:
Like my comment below, this should be improved by saying `just by checking the values that it takes and that it returns` or in other words the `input and the output of the function`, and eventually more formally `the domain and the codomain of the function`